### PR TITLE
feat(FEC-10057): move the plugin manager to kaltura player

### PR DESCRIPTION
**Pull request by yairans**
### Description of the Changes

* Add plugin management:
   - `PluginManager` and `BasePlugin`
   - plugin types and interfaces:
   `KPPluginsConfigObject `, `IMiddlewareProvider`, `IPlugin`and `IUIComponentsProvider `
 
* Add ads management:
   - `Ad` and `adBreak`
   - `AdsController` and `ControllerProvider`
   - ad types and interfaces:
   `KPAdBreakObject`, `KPAdPod`, `KPAdObject`, `KPAdvertisingConfigObject`, `IAdsController`, `IAdsControllerProvider` and `IAdsPluginController`
   - advertising and waterfalling management

* Add `KPUIComponent` and `IUIComponentsProvider`

* Add `PLAYBACK_START` and `PLAYBACK_ENDED` firing and handling

Solves FEC-10057

Related to https://github.com/kaltura/playkit-js/pull/463

### CheckLists

- [x] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [x] Docs have been updated

### CODE REVIEW

File: karma.conf.js [hunk 1]
```diff
@@ -9,19 +9,34 @@ const isMacOS = /^darwin/.test(process.platform);
...
 };
 
-module.exports = function(config) {
+const launchers = {
+  Chrome_browser: {
+    base: 'Chrome',
+    flags: ['--no-sandbox', '--autoplay-policy=no-user-gesture-required']
+  }
+};
+module.exports = function (config) {
```
Comment (dan-ziv) [inline karma.conf.js:22]: Add space

File: karma.conf.js [hunk 2]
```diff
@@ -9,19 +9,34 @@ const isMacOS = /^darwin/.test(process.platform);
...
+    customLaunchers: launchers,
+    browsers: ['Chrome_browser', 'Firefox'],
     concurrency: 1,
     singleRun: true,
     colors: true,
     frameworks: ['mocha'],
-    files: ['test/setup/karma.js'],
+    files: ['test/setup/karma.js',
+      {
+        pattern: 'test/src/assets/mov_bbb.mp4',
```
Comment (dan-ziv) [inline karma.conf.js:33]: assets shouldn't be under src/
put the under test folder directly

File: src/index.html [hunk 3]
```diff
@@ -6,18 +6,60 @@
   <title>Title</title>
   <link rel="stylesheet" type="text/css" href="./style.css"/>
   <script src="./kaltura-ovp-player.js" type="text/javascript"></script>
+  <script src="http://localhost:8082/playkit-vr.js" type="text/javascript"></script>
+  <script src="http://localhost:8083/playkit-ima.js" type="text/javascript"></script>
+<!--    <link rel="stylesheet" type="text/css" href="../dist/style.css"/>-->
```
Comment (dan-ziv) [inline src/index.html:11]: Remove comments

File: src/index.js [hunk 4]
```diff
@@ -21,6 +21,13 @@ import {getPlayers, getPlayer} from './proxy';
 import {cast} from './common/cast';
 // Import playlist
 import {playlist} from './common/playlist';
+// Import plugin framework
+import {BasePlugin} from './common/plugins/base-plugin';
```
Comment (dan-ziv) [inline src/index.js:25]: We didn't do it so far, but I would suggest to add index.js under each logic module (plugins, ads, etc...) and export all of its components from there. This is a good practice and very handy when you need to import multiple components from a certain module. You can then just do `export * from './common/plugins'`

File: src/kaltura-player.js [hunk 5]
```diff
@@ -78,7 +95,7 @@ class KalturaPlayer extends FakeEventTarget {
     const playerConfig = Utils.Object.copyDeep(mediaConfig);
     Utils.Object.mergeDeep(playerConfig.sources, this._localPlayer.config.sources);
     Utils.Object.mergeDeep(playerConfig.session, this._localPlayer.config.session);
-    Object.keys(this._localPlayer.config.plugins).forEach(name => {
+    Object.keys(this._localPlayer.config.plugins || {}).forEach(name => {
```
Comment (dan-ziv) [inline src/kaltura-player.js:98]: Why not saving the `options` on the `this` and access it by `this._options.plugins`?

File: karma.conf.js (hunk 1 above)
Comment (yairans) [reply karma.conf.js:23]: done

File: karma.conf.js (hunk 2 above)
Comment (yairans) [reply karma.conf.js]: done

File: src/index.html (hunk 3 above)
Comment (yairans) [reply src/index.html]: done

File: src/kaltura-player.js [hunk 6]
```diff
@@ -465,11 +496,23 @@ class KalturaPlayer extends FakeEventTarget {
...
-    return this._localPlayer.ads;
+    return this._adsController;
   }
 
   get plugins(): {[name: string]: BasePlugin} {
-    return this._localPlayer.plugins;
+    return this._pluginManager.getAll();
+  }
+
+  get uiComponents(): Array<PKUIComponent> {
```
Comment (Yuvalke) [inline src/kaltura-player.js:506]: does it still suppose to be an API? doesn't it suppose to be internal for UI?

File: src/index.js (hunk 4 above)
Comment (yairans) [reply src/index.js]: done

File: src/kaltura-player.js (hunk 6 above)
Comment (yairans) [reply src/kaltura-player.js]: done

File: src/kaltura-player.js [hunk 7]
```diff
@@ -173,9 +195,12 @@ class KalturaPlayer extends FakeEventTarget {
    */
   configure(config: Object = {}): void {
     config = supportLegacyOptions(config);
-    const configDoctionary = Utils.Object.mergeDeep({}, this.config, config);
-    evaluatePluginsConfig(config.plugins, configDoctionary);
-    this._localPlayer.configure(config);
+    const configDictionary = Utils.Object.mergeDeep({}, this.config, config);
+    evaluatePluginsConfig(config.plugins, configDictionary);
+    const localPlayerConfig = Utils.Object.mergeDeep({}, config);
+    delete localPlayerConfig.plugins;
```
Comment (Yuvalke) [inline src/kaltura-player.js:201]: why does it matter if you pass the plugins or not?
We are already passing the ui config which we don't need.
We can extract the config and pass the relevant config but I'm not sure if it's needed

File: src/kaltura-player.js [hunk 8]
```diff
@@ -535,6 +587,82 @@ class KalturaPlayer extends FakeEventTarget {
     return this._localPlayer.Error;
   }
 
+  _onChangeSourceStarted(): void {
+    this._configureOrLoadPlugins(this._pluginsConfig);
```
Comment (Yuvalke) [inline src/kaltura-player.js:591]: why don't you use it in setMedia instead?

File: src/kaltura-player.js (hunk 7 above)
Comment (yairans) [reply src/kaltura-player.js:201]: the plugins are not relevant for playkit. I would remove the UI as well

Comment (Yuvalke) [reply src/kaltura-player.js:201]: Let's make a helper function for that, you'll probably do the same here and in the constructor.

File: src/kaltura-player.js (hunk 8 above)
Comment (yairans) [reply src/kaltura-player.js:591]: I just mimic the [previous implement ](https://github.com/kaltura/playkit-js/pull/463/files#diff-29cfd888e32c875acb1a2dd113e516dbL486) in playkit

Comment (Yuvalke) [reply src/kaltura-player.js:591]: I see but I think it'll be cleaner to put it on `setMedia` instead if there isn't another reason

File: src/kaltura-player.js [hunk 9]
```diff
@@ -535,6 +587,82 @@ class KalturaPlayer extends FakeEventTarget {
...
+          }
+        } else {
+          delete plugins[name];
+        }
+      }
+    });
+    this._pluginsUiComponents = uiComponents;
+    middlewares.forEach(middleware => this._localPlayer.playbackMiddleware.use(middleware));
+    Utils.Object.mergeDeep(this._pluginsConfig, plugins);
+    this._maybeCreateAdsController();
```
Comment (Yuvalke) [inline src/kaltura-player.js:651]: why does this locate here?

File: src/kaltura-player.js [hunk 10]
```diff
@@ -36,19 +39,38 @@ class KalturaPlayer extends FakeEventTarget {
...
+    this._pluginsUiComponents = [];
+    this.configure({plugins});
     this._uiWrapper = new UIWrapper(this, options);
     this._provider = new Provider(options.provider, __VERSION__);
     this._playlistManager = new PlaylistManager(this, options);
     this._playlistManager.configure(options.playlist);
     Object.values(CoreEventType).forEach(coreEvent => this._eventManager.listen(this._localPlayer, coreEvent, e => this.dispatchEvent(e)));
+    this._eventManager.listen(this, CoreEventType.CHANGE_SOURCE_STARTED, () => this._onChangeSourceStarted());
+    this._eventManager.listen(this, CoreEventType.ENDED, () => this._onEnded());
+    this._eventManager.listen(this, CoreEventType.SOURCE_SELECTED, () => (this._sourceSelected = true));
```
Comment (OrenMe) [inline src/kaltura-player.js:73]: won't this be the same as setting it when setMedia is called or when sources are passed via constructor?
and actually, won't it be safer and more correct to pass via setMedia if sources are passed via constructor?

File: src/kaltura-player.js [hunk 11]
```diff
@@ -36,19 +39,38 @@ class KalturaPlayer extends FakeEventTarget {
...
     const noSourcesOptions = Utils.Object.mergeDeep({}, options, {sources: null});
+    delete noSourcesOptions.plugins;
     this._localPlayer = loadPlayer(noSourcesOptions);
     this._logger = getLogger('KalturaPlayer' + Utils.Generator.uniqueId(5));
+    this._reset = true;
+    this._pluginsConfig = {};
+    this._pluginEvents = {};
+    this._pluginManager = new PluginManager();
+    this._controllerProvider = new ControllerProvider(this._pluginManager);
+    this._pluginsUiComponents = [];
```
Comment (OrenMe) [inline src/kaltura-player.js:64]: I think that for clarity we should init all in the class members declaration above

File: src/kaltura-player.js (hunk 8 above)
Comment (yairans) [reply src/kaltura-player.js:611]: sources could be passed via `configure` on the config object. we can't rely on `setMedia` only

File: src/kaltura-player.js (hunk 9 above)
Comment (yairans) [reply src/kaltura-player.js:685]: after the plugin creation or update, it's the time to create the ads controller which uses the plugin manager

File: src/kaltura-player.js [hunk 12]
```diff
@@ -465,11 +504,23 @@ class KalturaPlayer extends FakeEventTarget {
...
+    return this._adsController;
   }
 
   get plugins(): {[name: string]: BasePlugin} {
-    return this._localPlayer.plugins;
+    return this._pluginManager.getAll();
+  }
+
+  get uiComponents(): Array<KPUIComponent> {
+    return [...this._pluginsUiComponents];
```
[the remark this answers is not in the document]
Comment (yairans) [reply src/kaltura-player.js:516]: the ui wrapper is created also by the [remote-control](https://github.com/kaltura/kaltura-player-js/blob/1e003e6bb97840eb55bbe91a1e5afc3353740b44/src/common/cast/remote-control.js#L117) 
and there we don't have the UI component on the config to pass to the UI wrapper constructor

File: src/kaltura-player.js [hunk 13]
```diff
@@ -465,11 +504,23 @@ class KalturaPlayer extends FakeEventTarget {
   }
 
   get ads(): ?AdsController {
-    return this._localPlayer.ads;
+    return this._adsController;
```
[the remark this answers is not in the document]
Comment (yairans) [reply src/kaltura-player.js:508]: no. this api returns `AdsController`

File: src/kaltura-player.js [hunk 14]
```diff
@@ -490,6 +541,7 @@ class KalturaPlayer extends FakeEventTarget {
       Core: CoreEventType,
       Playlist: PlaylistEventType,
       UI: UIEventType,
+      Plugins: this._pluginEvents,
```
[the remark this answers is not in the document]
Comment (yairans) [reply src/kaltura-player.js:545]: removed.

File: src/kaltura-player.js [hunk 15]
```diff
@@ -535,6 +587,83 @@ class KalturaPlayer extends FakeEventTarget {
...
+    // Make sure the all ENDED listeners have been invoked
+    setTimeout(() => {
+      if (this._adsController && !this._adsController.allAdsCompleted) {
+        this._eventManager.listenOnce(this._adsController, AdEventType.ALL_ADS_COMPLETED, () => {
+          this.dispatchEvent(new FakeEvent(CoreEventType.PLAYBACK_ENDED));
+        });
+      } else {
+        this.dispatchEvent(new FakeEvent(CoreEventType.PLAYBACK_ENDED));
+      }
+    });
```
[the remark this answers is not in the document]
Comment (yairans) [reply src/kaltura-player.js:627]: this is for when the player ended listener is registered before the ad controller listener.
like by:
```
    var kalturaPlayer = KalturaPlayer.setup(config);
    kalturaPlayer.configure({
      plugins: {
         ima: {
            ...
         }
      }
    });
```
I pretty sure I had to do the same in playkit. anyway it's safe.

File: src/kaltura-player.js (hunk 10 above)
Comment (yairans) [reply src/kaltura-player.js]: for 1. `SOURCE_SELECTED` is fired only when there is actually source to play (hls, dash or progressive). i don't want to add this validation also here.
for 2. how it's related to this change

File: src/kaltura-player.js (hunk 8 above)
Comment (OrenMe) [reply src/kaltura-player.js:604]: so why not use setMedia always, also internally?
basically this is can almost be considered a bug as we have logic around adding sources to player which is enforced in setmedia.
the suggestion here is to also use setMedia if the sources are provided in the config passed to setup call.

File: src/kaltura-player.js (hunk 11 above)
Comment (yairans) [reply src/kaltura-player.js]: done

File: src/kaltura-player.js (hunk 8 above)
Comment (yairans) [reply src/kaltura-player.js:604]: agreed but I said above this is out of scope of this change so I prefer to do it separately

File: src/kaltura-player.js (hunk 5 above)
Comment (dan-ziv) [reply src/kaltura-player.js]: @yairans ?

File: src/kaltura-player.js [hunk 16]
```diff
@@ -586,6 +595,7 @@ class KalturaPlayer extends FakeEventTarget {
     this._eventManager.listen(this, CoreEventType.ENDED, () => this._onEnded());
     this._eventManager.listen(this, CoreEventType.FIRST_PLAY, () => (this._firstPlay = false));
     this._eventManager.listen(this, CoreEventType.SOURCE_SELECTED, () => (this._sourceSelected = true));
+    this._eventManager.listen(this, CoreEventType.PLAYBACK_ENDED, this._onPlaybackEnded.bind(this));
```
Comment (Yuvalke) [inline src/kaltura-player.js:598]: I think it'll be better to keep the same syntax. 
arrow function or bind function

File: dist/translations/de.i18n.json [hunk 17]
```diff
@@ -1 +1 @@
...
...ault_error":"Etwas ist schiefgegangen","default_session_text":"Sitzungs-ID","retry":"Erneut versuchen"},"ads":{"ad_notice":"Anzeige","learn_more":"Weitere Informationen","skip_ad":"Werbung überspringen","skip_in":"Überspringen in"},"cvaa":{"title":"Erweiterte Einstellungen für Untertitel","sample_caption_tag":"Beispiel","sample_custom_caption_tag":"Benutzerdefinierte Untertitel","set_custom_caption":"Benutzerdefinierte Untertitel auswählen","edit_caption":"Untertitel bearbeiten","size_label":"Größe","font_color_label":"Schriftfarbe","font_family_label":"Schriftart","font_style_label":"Schriftstil","font_opacity_label":"Deckkraft","background_color_label":"Hintergrundfarbe","background_opacity_label":"Deckkraft Hintergrund","apply":"Übernehmen","caption_preview":"Untertitelvorschau"},"cast":{"play_on_tv":"Über Fernseher abspielen","disconnect_from_tv":"Verbindung zum Fernseher trennen","status":{"connecting_to":"Verbinden mit","connected_to":"Verbunden mit","playing_on":"Wiedergabe auf"}},"playlist":{"prev":"Zurück","next":"Weiter","up_next":"Als Nächstes","cancel":"Abbrechen"},"pictureInPicture":{"overlay_text":"Wiedergabe im Bild-im-Bild-Modus","overlay_button":"Hier abspielen"}}}
```
Comment (Yuvalke) [inline dist/translations/de.i18n.json:1]: need to remember to remove dist files before merge.

File: src/kaltura-player.js (hunk 5 above)
Comment (yairans) [reply src/kaltura-player.js]: this code is outdated @dan-ziv

File: src/kaltura-player.js (hunk 16 above)
Comment (OrenMe) [reply src/kaltura-player.js:598]: @yairans please fix

Comment (yairans) [reply src/kaltura-player.js:600]: already fixed

File: src/kaltura-player.js (hunk 7 above)
Comment (Yuvalke) [reply src/kaltura-player.js:195]: @yairans you didn't remove the irrelevant config

File: src/kaltura-player.js [hunk 18]
```diff
@@ -36,19 +39,38 @@ class KalturaPlayer extends FakeEventTarget {
...
+  _reset: boolean;
+  _sourceSelected: boolean;
 
   constructor(options: KPOptionsObject) {
     super();
     this._eventManager = new EventManager();
-    const {sources} = options;
+    const {sources, plugins} = options;
     const noSourcesOptions = Utils.Object.mergeDeep({}, options, {sources: null});
+    delete noSourcesOptions.plugins;
```
[the remark this answers is not in the document]
Comment (Yuvalke) [reply src/kaltura-player.js:56]: @yairans need to remove irrelevant config as well

File: flow-typed/types/kaltura-player-options.js [hunk 19]
```diff
@@ -7,7 +7,8 @@ declare type KPOptionsObject = {
   disableUserCache?: boolean,
   playback?: PKPlaybackConfigObject,
   sources?: PKSourcesConfigObject,
-  plugins?: PKPluginsConfigObject,
+  plugins: KPPluginsConfigObject,
```
Comment (RoyBregman) [inline flow-typed/types/kaltura-player-options.js:10]: it is not optional anymore?

Comment (yairans) [reply flow-typed/types/kaltura-player-options.js:10]: it's {} by default

Comment (dan-ziv) [reply flow-typed/types/kaltura-player-options.js:10]: We are initialise it internally, but from the app/user perspective this is optionally.

File: src/kaltura-player.js [hunk 20]
```diff
@@ -14,41 +14,55 @@ import {CastEventType} from './common/cast/cast-event-type';
...
     super();
-    this._eventManager = new EventManager();
-    const {sources} = options;
+    const {sources, plugins} = options;
     const noSourcesOptions = Utils.Object.mergeDeep({}, options, {sources: null});
+    delete noSourcesOptions.plugins;
     this._localPlayer = loadPlayer(noSourcesOptions);
-    this._logger = getLogger('KalturaPlayer' + Utils.Generator.uniqueId(5));
+    this._controllerProvider = new ControllerProvider(this._pluginManager);
+    this.configure({plugins});
```
Comment (Yuvalke) [inline src/kaltura-player.js:59]: why do we call configure twice? this.confugure and this._localPlayer.configure calls this._localPlayer.configure

Comment (yairans) [reply src/kaltura-player.js:59]: the first configs the plugins the second configs the sources

**Pull request merged**